src/contracts/caching_log_query.rs

```diff
@@ -76,29 +93,72 @@ impl CachingLogQuery {
                 yield serde_json::from_str(&log).map_err(Error::Parse)?;
             }
 
-            let filter = self.filter.clone().from_block(max(
-                last_block.db + 1,
-                self.filter.get_from_block().unwrap_or_default(),
-            ));
-            let mut stream = self.provider
-                .get_logs_paginated(&filter, self.page_size as u64);
-
-            while let Some(log) = stream.next().await {
-                let log = log.map_err(Error::LoadLogs)?;
-                if self.is_confirmed(&log, last_block) {
-                    let raw_log = serde_json::to_string(&log).map_err(Error::Serialize)?;
-                    self.cache_log(raw_log, &log).await?;
+            let mut retry_stats = RetryStatus::new(self.start_page_size, self.min_page_size, self.max_backoff_time);
+
+            'restart: loop {
+                let filter = self.filter.clone().from_block(max(
+                    max(retry_stats.last_block + 1, last_block.db + 1),
+                    self.filter.get_from_block().unwrap_or_default(),
+                ));
+
+                info!(
+                    page_size = retry_stats.page_size,
+                    from_block = filter.get_from_block().unwrap_or_default().as_u64(),
+                    "Reading MemberAdded events from chains"
+                );
+
+                let mut stream = self.provider
+                    .get_logs_paginated(&filter, retry_stats.page_size);
+
+                while let Some(log) = stream.next().await {
+                    let log = match self.handle_retriable_err(log, &mut retry_stats).await {
+                        RetriableResult::Ok(log) => log,
+                        RetriableResult::Restart => continue 'restart,
+                        RetriableResult::Err(e) => {
+                            // This seems to be the only way to return errors in the try_stream! macro
+                            Err(e)?;
+                            break 'restart;
+                        }
+                    };
+
+                    // If we need to decrease page size later, don't process same blocks twice
+                    retry_stats.update_last_block(log.block_number);
+
+                    if self.is_confirmed(&log, last_block) {
+                        let raw_log = serde_json::to_string(&log).map_err(Error::Serialize)?;
+                        self.cache_log(raw_log, &log).await?;
+                    }
+
+                    yield log;
                 }
-                yield log;
+
+                // We've iterated over all events, we can kill the restart loop
+                break;
             }
         }
     }
 
+    async fn handle_retriable_err(
+        &self,
+        log: Result<Log, LogQueryError<ProviderError>>,
+        stats: &mut RetryStatus,
+    ) -> RetriableResult<Log, Error<ProviderError>> {
+        match log {
+            Err(e) if e.to_string().contains("Query timeout exceeded") => {
```

Comment:
Alchemy's status code is hidden behind
```
#[error(transparent)]
JsonRpcClientError(#[from] Box<dyn std::error::Error + Send + Sync>),
```
which makes it really difficult to match without `string.contains(...)` method.